### PR TITLE
PYTHON-3758 Support overflow integers in fallback_encoder.

### PR DESCRIPTION
**Pull request by stephan-hof**
Hi,

I'm regularly dealing with 64-bit *unsigned* integers.
Those range from `[0, (2**64) - 1]`. So they are outside of the 64-bit *signed* integers from  [BSONs](https://bsonspec.org/spec.html) `"\x12" e_name int64 64-bit integer`.

Up to now pymongo rejects those immediately with an OverflowError so the application must deal with them.
For us this is pretty cumbersome, since we have a pretty complex structure and finding those large numbers would mean to 'traverse' through the whole document before giving it to pymongo.

Performance and code wise this is not very ideal so my thinking was:
To convert the document to BSON, pymongo must traverse the document anyway.
So lets handle this 'big integer' like any other unknown type and call the `fallback_encoder` (if given). 
Then let the fallback_encoder deal with it. This is now up to me, like it is already with unknown types.

In other words: If pymongo/bson detects an integer outside its supported range the `fallback_encoder` gets called.
If no `fallback_encoder` is given or if `fallback_encoder` gives back a 'big integer' again the usual `OverflowError` is thrown.

### CODE REVIEW

**Review by blink1073 — CHANGES_REQUESTED** [review]

File: bson/__init__.py [hunk 1]
```diff
@@ -896,12 +896,18 @@ def _name_value_to_bson(
...
+
+    was_integer_overflow = False
+
     # First see if the type is already cached. KeyError will only ever
     # happen once per subtype.
     try:
         return _ENCODERS[type(value)](name, value, check_keys, opts)  # type: ignore
     except KeyError:
         pass
+    except IntegerOverflow:
```
[the remark this answers is not in the document]
Comment (stephan-hof) [reply bson/__init__.py]: Many thanks for giving this pull-requestion attention. Your suggestion makes sense to me and already pushed a commit to address it. + merged latest master into the branch as well.

How do you handle the 'resolve conversation' ? Should I press it or is this something the reviewer does after accepting the change ?

[the remark this answers is not in the document]
Comment (blink1073) [reply bson/__init__.py]: The reviewer does it.  Done!

**Review by blink1073 — APPROVED** [review]
LGTM, thank you!

File: bson/__init__.py [hunk 2]
```diff
@@ -896,12 +896,20 @@ def _name_value_to_bson(
...
+    was_integer_overflow = False
+
     # First see if the type is already cached. KeyError will only ever
     # happen once per subtype.
     try:
         return _ENCODERS[type(value)](name, value, check_keys, opts)  # type: ignore
     except KeyError:
         pass
+    except OverflowError as error:
+        if error.args != ("BSON can only handle up to 8-byte ints",):
```
Comment (ShaneHarvey) [inline bson/__init__.py:909]: Could you replace this with `if not isinstance(value, int):`? This way the fallback encoder won't be called multiple times for one OverflowError in this case:
```
>>> bson.encode({'a' {'b': 'c': 2<<65}}, codec_options=CodecOptions(...fallback_encoder=lambda val: val)))

```

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Feel free to add yourself to `doc/contributors.rst` if you like!

File: test/test_custom_types.py [hunk 3]
```diff
@@ -274,6 +274,22 @@ def fallback_encoder(value):
         with self.assertRaises(TypeError):
             encode(document, codec_options=codecopts)
 
+    def test_call_only_once_for_not_handled_big_integers(self):
```
Comment (ShaneHarvey) [inline test/test_custom_types.py:277]: Nice test!

**Review by ShaneHarvey — APPROVED** [review]
Thanks @stephan-hof, LGTM!

**Comment by stephan-hof** [comment]
Just wanted to say: Many thanks for taking this serious and giving me a great contribution experience.
